### PR TITLE
add benchmark test for ipam

### PR DESCRIPTION
**Pull request by changluyi**
- [ ] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

#### What type of this PR
Examples of user facing changes:
- Features
- Bug fixes
- Docs
- Tests

add serial and Parallel test cases
serial:
1.  add/del 10000  ipv4/ipv6/dual subnets
2.  allocate/free 10000 ipv4/ipv6/dual pod ips

Parallel :
1. add/del 1000 ipv4/ipv6/dual subnets with 10 worker
2. allocate/free 4000 ipv4/ipv6/dual pod ips with 10 worker

result:
```
[root@localhost kube-ovn]# make ipam-bench
go test -timeout 30m -bench='^BenchmarkIPAM' -benchtime=10000x test/unittest/ipam_bench/ipam_test.go -args -logtostderr=false
goos: linux
goarch: amd64
cpu: AMD Ryzen 7 5800H with Radeon Graphics
BenchmarkIPAMSerialIPv4AddSubnet-8         10000             27260 ns/op
BenchmarkIPAMSerialIPv4DelSubnet-8         10000             10989 ns/op
BenchmarkIPAMSerialIPv4AllocAddr-8         10000             18806 ns/op
BenchmarkIPAMSerialIPv4FreeAddr-8          10000             18368 ns/op
BenchmarkIPAMSerialIPv6AddSubnet-8         10000             21228 ns/op
BenchmarkIPAMSerialIPv6DelSubnet-8         10000             10963 ns/op
BenchmarkIPAMSerialIPv6AllocAddr-8         10000             21143 ns/op
BenchmarkIPAMSerialIPv6FreeAddr-8          10000             19123 ns/op
BenchmarkIPAMSerialDualAddSubnet-8         10000             32336 ns/op
BenchmarkIPAMSerialDualDelSubnet-8         10000             10665 ns/op
BenchmarkIPAMSerialDualAllocAddr-8         10000             31298 ns/op
BenchmarkIPAMSerialDualFreeAddr-8          10000             35458 ns/op
BenchmarkIPAMRandomIPv4AllocAddr-8         10000           3327854 ns/op
BenchmarkIPAMRandomIPv4FreeAddr-8          10000           8881645 ns/op
BenchmarkIPAMRandomIPv6AllocAddr-8         10000           3631844 ns/op
BenchmarkIPAMRandomIPv6FreeAddr-8          10000           9956444 ns/op
BenchmarkIPAMRandomDualAllocAddr-8         10000           6332809 ns/op
BenchmarkIPAMRandomDualFreeAddr-8          10000          18652224 ns/op
PASS
ok      command-line-arguments  649.970s
go test -timeout 90m -bench='^BenchmarkParallelIPAM' -benchtime=10x test/unittest/ipam_bench/ipam_test.go -args -logtostderr=false
goos: linux
goarch: amd64
cpu: AMD Ryzen 7 5800H with Radeon Graphics
BenchmarkParallelIPAMIPv4AddDel1000Subnet-8           10          26068826 ns/op
BenchmarkParallelIPAMIPv4AllocFree3000Addr-8          10        21639633405 ns/op
BenchmarkParallelIPAMIPv6AddDel1000Subnet-8           10          21706920 ns/op
BenchmarkParallelIPAMIPv6AllocFree3000Addr-8          10        32859503226 ns/op
BenchmarkParallelIPAMDualAddDel1000Subnet-8           10          30747915 ns/op
BenchmarkParallelIPAMDualAllocFree3000Addr-8          10        104593247919 ns/op
PASS
ok      command-line-arguments  1623.125s

```

when delete 10w pod', the latency is very high, with below

```
[root@localhost kube-ovn]# go test -timeout 6h -bench='^BenchmarkIPAMRandomIPv4AllocAddrPerformanceTrace' -benchtime=100000x test/unittest/ipam_bench/ipam_test.go -args -logtostderr=false
goos: linux
goarch: amd64
cpu: AMD Ryzen 7 5800H with Radeon Graphics
BenchmarkIPAMRandomIPv4AllocAddrPerformanceTrace-8      0 to 9999 pods spent time 32s
10000 to 19999 pods spent time 110s
20000 to 29999 pods spent time 164s
30000 to 39999 pods spent time 267s
40000 to 49999 pods spent time 336s
50000 to 59999 pods spent time 412s
60000 to 69999 pods spent time 514s
70000 to 79999 pods spent time 511s
80000 to 89999 pods spent time 2072s
90000 to 99999 pods spent time 696s
delete 0 to 9999 pods spent time 97s
delete 10000 to 19999 pods spent time 313s
delete 20000 to 29999 pods spent time 525s
delete 30000 to 39999 pods spent time 3197s
delete 40000 to 49999 pods spent time 1019s
delete 50000 to 59999 pods spent time 1844s
delete 60000 to 69999 pods spent time 3287s
```

#### Which issue(s) this PR fixes:
Fixes #(issue-number)

### CODE REVIEW

File: test/unittest/ipam_bench/ipam_test.go [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+
+import (
+	"fmt"
+	kubeovnv1 "github.com/kubeovn/kube-ovn/pkg/apis/kubeovn/v1"
+	"github.com/kubeovn/kube-ovn/pkg/ipam"
+
+	"k8s.io/klog/v2"
+	"math/rand"
+	"testing"
+)
```
Comment (zhangzujian) [inline test/unittest/ipam_bench/ipam_test.go:11]:
```suggestion
import (
	"fmt"
	"math/rand"
	"testing"

	"k8s.io/klog/v2"

	kubeovnv1 "github.com/kubeovn/kube-ovn/pkg/apis/kubeovn/v1"
	"github.com/kubeovn/kube-ovn/pkg/ipam"
)
```

**Comment by zhangzujian** [comment]
Benchmark results should be attached.

**Comment by zhangzujian** [comment]
We should avoid the massive logs in benchmark ut:

```txt
I1205 02:17:27.082457    2717 subnet.go:455] release v4 10.0.17.31 mac 00:00:00:8E:B7:2D for pod4381, add ip to released list
I1205 02:17:27.082473    2717 subnet.go:455] release v4 10.0.17.32 mac 00:00:00:50:7A:8B for pod4382, add ip to released list
I1205 02:17:27.082503    2717 subnet.go:455] release v4 10.0.17.33 mac 00:00:00:D2:7D:0D for pod4383, add ip to released list
I1205 02:17:27.082537    2717 subnet.go:455] release v4 10.0.17.34 mac 00:00:00:A5:5A:F1 for pod4384, add ip to released list
I1205 02:17:27.082552    2717 subnet.go:455] release v4 10.0.17.35 mac 00:00:00:39:53:6D for pod4385, add ip to released list
I1205 02:17:27.082567    2717 subnet.go:455] release v4 10.0.17.36 mac 00:00:00:C2:2B:BE for pod4386, add ip to released list
I1205 02:17:27.082599    2717 subnet.go:455] release v4 10.0.17.37 mac 00:00:00:9A:41:71 for pod4387, add ip to released list
I1205 02:17:27.082614    2717 subnet.go:455] release v4 10.0.17.38 mac 00:00:00:87:3A:7B for pod4388, add ip to released list
```

File: test/unittest/ipam_bench/ipam_test.go [hunk 2]
```diff
@@ -0,0 +1,423 @@
+package ipam
```
Comment (zhangzujian) [inline test/unittest/ipam_bench/ipam_test.go:1]:
```suggestion
package ipam_bench
```

File: test/unittest/ipam_bench/ipam_test.go [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+	testing.Init()
+	klog.InitFlags(nil)
+	flag.Parse()
+}
+
+func BenchmarkIPAMIPv4AddSubnet(b *testing.B) {
+	im := ipam.NewIPAM()
+	for n := 0; n < b.N; n++ {
+		if ok := addIPAMSubnet(im, n, kubeovnv1.ProtocolIPv4); !ok {
+			klog.Fatalf("ERROR: add v4 subnet with index %d ", n)
```
Comment (zhangzujian) [inline test/unittest/ipam_bench/ipam_test.go:59]: Use `b.Errorf()` instead.

File: test/unittest/ipam_bench/ipam_test.go [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+func getDefaultSubnetRandomIps(protocol string, ipCount int) []string {
+
+	var ips []string
+	var newip string
+	for n := 0; n < ipCount; n++ {
+		b := make([]byte, 3)
+		_, err := rand.Read(b)
+		if err != nil {
+			klog.Fatal("generate random error: %v", err)
+		}
```
Comment (zhangzujian) [inline test/unittest/ipam_bench/ipam_test.go:339]:
```suggestion
		if _, err := rand.Read(b); err != nil {
			b.Errorf("generate random error: %v", err)
		}
```

File: test/unittest/ipam_bench/ipam_test.go [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+}
+
+func delSubnetCapacity(b *testing.B, im *ipam.IPAM) {
+	for n := 0; n < b.N; n++ {
+		delIPAMSubnet(im, n)
+	}
+}
+
+func addSerailAddrCapacity(b *testing.B, im *ipam.IPAM, protocol string) {
+	subnetName, CIDR, Gw, ExcludeIPs := getDefaultSubnetParam(protocol)
```
Comment (zhangzujian) [inline test/unittest/ipam_bench/ipam_test.go:172]:
```suggestion
	subnetName, cidr, gw, excludeIPs := getDefaultSubnetParam(protocol)
```

File: test/unittest/ipam_bench/ipam_test.go [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+			return false
+		}
+	} else if protocol == kubeovnv1.ProtocolIPv6 {
+		if err := im.AddOrUpdateSubnet(subnetName, ipv6CIDR, v6Gw, ipv6ExcludeIPs); err != nil {
+			klog.Fatal("ERROR: add subnet with ipv6 cidr %s, with index %d err %v ", ipv6CIDR, index, err)
+			return false
+		}
+	} else if protocol == kubeovnv1.ProtocolDual {
+		if err := im.AddOrUpdateSubnet(subnetName, dualCIDR, dualGw, dualExcludeIPs); err != nil {
+			klog.Fatal("ERROR: add subnet with dual cidr %s, with index %d err %v ", dualCIDR, index, err)
```
Comment (zhangzujian) [inline test/unittest/ipam_bench/ipam_test.go:243]: Use `b.Errorf()`.

File: test/unittest/ipam_bench/ipam_test.go [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+		}
+		if protocol == kubeovnv1.ProtocolIPv4 {
+			newip = fmt.Sprintf("10.%d.%d.%d", b[0], b[1], b[2])
+		} else if protocol == kubeovnv1.ProtocolIPv6 {
+			newip = fmt.Sprintf("fd00::00%02x:%02x%02x", b[0], b[1], b[2])
+		} else {
+			newip = fmt.Sprintf("10.%d.%d.%d,fd00::00%02x:%02x%02x",
+				b[0], b[1], b[2], b[0], b[1], b[2])
+		}
+		if ok := slices.Contains(ips, newip); ok {
```
Comment (zhangzujian) [inline test/unittest/ipam_bench/ipam_test.go:348]:
```suggestion
		if slices.Contains(ips, newip) {
```

File: test/unittest/ipam_bench/ipam_test.go [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+
+import (
+	"flag"
+	"fmt"
+	"k8s.io/utils/strings/slices"
+	"math/big"
+	"testing"
+
+	"crypto/rand"
+	"k8s.io/klog/v2"
```
Comment (zhangzujian) [inline test/unittest/ipam_bench/ipam_test.go:11]:
```suggestion
	"crypto/rand"
	"flag"
	"fmt"
	"math/big"
	"testing"

	"k8s.io/klog/v2"
	"k8s.io/utils/strings/slices"
```

Comment (zhangzujian) [reply test/unittest/ipam_bench/ipam_test.go:11]: Why use `crypto/rand` instead of `math/rand`?

Comment (zhangzujian) [reply test/unittest/ipam_bench/ipam_test.go:11]: 1. What about the performance difference? AFAIK, `crypto/rand` is used mainly for security.
2. Before using `math/rand`, you need to initialize the random source:

```go
func init() {
	rand.Seed(time.Now().UnixNano())
}
```

Comment (changluyi) [reply test/unittest/ipam_bench/ipam_test.go]: 1.  no significant difference and I see crypto/rand is used in pkg/util/net.go  such as GenerateMac
2. got

File: test/unittest/ipam_bench/ipam_test.go [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+
+func (stringSet *StringSet) Add(item string) bool {
+	if _, ok := stringSet.StringMap[item]; ok {
+		return false
+	}
+	stringSet.StringMap[item] = struct{}{}
+	return true
+}
+
+func (stringSet *StringSet) Pop() (string, bool) {
```
Comment (zhangzujian) [inline test/unittest/ipam_bench/ipam_test.go:388]:
```suggestion
func (s *StringSet) Pop() (string, bool) {
```

File: test/unittest/ipam_bench/ipam_test.go [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+		switch protocol {
+		case kubeovnv1.ProtocolIPv4:
+			newIp = fmt.Sprintf("10.%d.%d.%d", bytes[0], bytes[1], bytes[2])
+		case kubeovnv1.ProtocolIPv6:
+			newIp = fmt.Sprintf("fd00::00%02x:%02x%02x", bytes[0], bytes[1], bytes[2])
+		case kubeovnv1.ProtocolDual:
+			newIp = fmt.Sprintf("10.%d.%d.%d,fd00::00%02x:%02x%02x",
+				bytes[0], bytes[1], bytes[2], bytes[0], bytes[1], bytes[2])
+		}
+		ipSet.Add(newIp)
```
Comment (zhangzujian) [inline test/unittest/ipam_bench/ipam_test.go:359]: We should check the result.

File: test/unittest/ipam_bench/ipam_test.go (hunk 9 above)
Comment (changluyi) [reply test/unittest/ipam_bench/ipam_test.go:388]: ok

**Pull request merged**